internal/retry/sizer.go

```diff
@@ -0,0 +1,87 @@
+package retry
+
+// SizerCheckFunc allows the creator of a Sizer to specify arbitrarily
+// complex checks of the error condition to determine if the size
+// needs to be retried.
+type SizerCheckFunc func(error) bool
+
+// Sizer is used to implement 'resize loops' that hides the complexity
+// of the sizing and error checking away from most of the application.
+//
+// At the end of each iteration call Update or UpdateWants to inform
+// the sizer of the result of the action(s) taken during the iteration.
+// If the error is nil the update functions will return nil and halt
+// the iteration. If the error is non nil but returns false from
+// the check function the same error is returned and iteration will halt.
+// If the check function returns true iteration continues but nil
+// will be returned from the update function, effectively clearing
+// the error state, *unless* the current size now exceeds maxSize.
+type Sizer struct {
+	size    int
+	maxSize int
+	ready   bool
+	again   SizerCheckFunc
+}
+
+// NewSizer returns a new Sizer ready for use.
+// The Sizer will attempt a retry on conditions where the callback
+// function f returns true.
+func NewSizer(startSize, maxSize int, f SizerCheckFunc) *Sizer {
+	return &Sizer{
+		ready:   true,
+		size:    startSize,
+		maxSize: maxSize,
+		again:   f,
+	}
+}
+
+// NewSizerEV returns a new Sizer ready for use that only checks
+// for the exact error value specified.
+func NewSizerEV(startSize, maxSize int, retryOn error) *Sizer {
```

Comment:
I wonder if this really adds a benefit. The SizerCheckFunc can easily see if the returned errno indicates a retry. The added internal/private errors do not make the code easier to understand imho.

Reply:
I like the generic nature of the check-func but for many simple use cases it is more effort than needed and thus I added the new function to check for only a specific value.
I purposefully avoided importing "C" in the retry package. But I would be open to having a private helper function in the rbd/rados/cephfs packages to "bake in" checking for ERANGE/etc if you prefer that approach.

Reply:
I think I prefer the approach to have a private function in the package that calls `NewSizer`. `NewSizerEV` can then be dropped.